pkg/expression/contextstatic/evalctx.go

```diff
@@ -0,0 +1,362 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package contextstatic
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pingcap/tidb/pkg/errctx"
+	exprctx "github.com/pingcap/tidb/pkg/expression/context"
+	"github.com/pingcap/tidb/pkg/expression/contextopt"
+	"github.com/pingcap/tidb/pkg/parser/mysql"
+	"github.com/pingcap/tidb/pkg/sessionctx/variable"
+	"github.com/pingcap/tidb/pkg/types"
+	contextutil "github.com/pingcap/tidb/pkg/util/context"
+	"github.com/pingcap/tidb/pkg/util/intest"
+)
+
+// StaticEvalContext should implement `EvalContext`
+var _ exprctx.EvalContext = &StaticEvalContext{}
+
+type timeOnce struct {
+	sync.Mutex
+	time   atomic.Pointer[time.Time]
+	timeFn func() (time.Time, error)
+}
+
+func (t *timeOnce) getTime(loc *time.Location) (tm time.Time, err error) {
+	if p := t.time.Load(); p != nil {
+		return *p, nil
+	}
+
+	t.Lock()
+	defer t.Unlock()
+
+	if p := t.time.Load(); p != nil {
+		return *p, nil
+	}
+
+	if fn := t.timeFn; fn != nil {
+		tm, err = fn()
+	} else {
+		tm, err = time.Now(), nil
+	}
+
+	if err != nil {
+		return
+	}
+
+	tm = tm.In(loc)
+	t.time.Store(&tm)
+	return
+}
+
+// staticEvalCtxState is the internal state for `StaticEvalContext`.
+// We make it as a standalone private struct here to make sure `StaticEvalCtxOption` can only be called in constructor.
+type staticEvalCtxState struct {
```

Comment:
Can you give some explaination about why the `staticEvalCtxState` is named by `static`?